File: day03/day03p2.py
import re
import sys

print(
    sum(
        int(n1) * int(n2)
        for n1, n2 in re.findall(
            r'mul\((\d+),(\d+)\)',
            re.sub(
                r"don't\(\).*?(do\(\)|$)", '', sys.stdin.read().replace('\n', ''), count=0
            ),
        )
    )
)

File: day03/Day03P2.kt
private val MUL = """mul\((\d+),(\d+)\)""".toRegex()
private val DO_NOT = """don't\(\).*?(do\(\)|$)""".toRegex()

fun main() {
  val contents = generateSequence(::readLine).joinToString("")
  println(
      MUL.findAll(contents.replace(DO_NOT, ""))
      .map { it.groupValues.drop(1).map(String::toInt) }
      .map { (n1, n2) -> n1 * n2 }
      .sum()
  )
}
